### PR TITLE
refactor: ir.TCPListener to support one to many listener to routes 

### LINKED ISSUE #1635: Refactor Xds Listener for TLSRoutes
*Description*:

Today, the TLSRoute name is incorporated into the Xds Listener name so when 1 of many TLS Routes gets deleted, the Xds Listener gets recreated which is not great.

Instead, 

* Create 1 IR TCP Listener per Gateway Listener
* Map TLSRoutes context into the same IR TCP Listener, maybe by creating an array of IR TCP Routes
* Create a unique Filter Chain match for every IR TCP Route, using SNIs as the unique differentiator

thanks @dboslee for catching this 

Relates to https://github.com/envoyproxy/gateway/issues/1634#issuecomment-1626357641

### PR DESCRIPTION
**Pull request by dboslee**
**What this PR does / why we need it**:
This change refactors the ir.TCPListener to be named directly after the corresponding gateway listener. The tcp/tls routes map to TCPRoutes attached to the TCPListener.

**Which issue(s) this PR fixes**:
Fixes #1635

### CODE REVIEW

**Comment by arkodg** [comment]
thanks for fixing this, heads up, there are 2 PRs approved and ready to be merged #1702 #1788 that might require rebasing here

File: internal/gatewayapi/helpers.go [hunk 1]
```diff
@@ -413,15 +413,15 @@ func irStringKey(gatewayNs, gatewayName string) string {
 	return fmt.Sprintf("%s/%s", gatewayNs, gatewayName)
 }
 
-func irHTTPListenerName(listener *ListenerContext) string {
+func irListenerName(listener *ListenerContext) string {
 	return fmt.Sprintf("%s/%s/%s", listener.gateway.Namespace, listener.gateway.Name, listener.Name)
 }
 
-func irTLSListenerName(listener *ListenerContext, tlsRoute *TLSRouteContext) string {
+func irTLSRouteName(listener *ListenerContext, tlsRoute *TLSRouteContext) string {
```
Comment (arkodg) [inline internal/gatewayapi/helpers.go:420]: does `listener` need to be part of the name, can this helper look more like `irRouteName` w/o the the `matchIdx`

File: internal/gatewayapi/helpers.go [hunk 2]
```diff
@@ -413,15 +413,15 @@ func irStringKey(gatewayNs, gatewayName string) string {
...
 	return fmt.Sprintf("%s/%s/%s/%s", listener.gateway.Namespace, listener.gateway.Name, listener.Name, tlsRoute.Name)
 }
 
-func irTCPListenerName(listener *ListenerContext, tcpRoute *TCPRouteContext) string {
+func irTCPRouteName(listener *ListenerContext, tcpRoute *TCPRouteContext) string {
```
Comment (arkodg) [inline internal/gatewayapi/helpers.go:424]: does `listener` need to be part of the name, can this helper look more like `irRouteName` w/o the the `matchIdx`

File: internal/gatewayapi/listener.go [hunk 3]
```diff
@@ -119,6 +119,13 @@ func (t *Translator) ProcessListeners(gateways []*GatewayContext, xdsIR XdsIRMap
 					irListener.Hostnames = append(irListener.Hostnames, "*")
 				}
 				gwXdsIR.HTTP = append(gwXdsIR.HTTP, irListener)
+			case v1beta1.TCPProtocolType, v1beta1.TLSProtocolType:
```
Comment (arkodg) [inline internal/gatewayapi/listener.go:122]: dont we need to set `irListener.Hostnames` for `TLS` ?
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.Listener

File: internal/ir/xds.go [hunk 4]
```diff
@@ -679,7 +679,16 @@ type TCPListener struct {
 	Address string `json:"address" yaml:"address"`
 	// Port on which the service can be expected to be accessed by clients.
 	Port uint32 `json:"port" yaml:"port"`
-	// TLS holds information for configuring TLS on a listener
+	// Routes associated with TCP traffic to the listener.
+	Routes []*TCPRoute `json:"routes" yaml:"routes"`
```
Comment (arkodg) [inline internal/ir/xds.go:683]:
```suggestion
	Routes []*TCPRoute `json:"routes,omitempty" yaml:"routes,omitempty"`
```

**Comment by arkodg** [comment]
thanks @dboslee, looks mostly good, added some comments

**Comment by dboslee** [comment]
Was OOO, I will circle back to this in the next few days.

**Comment by arkodg** [comment]
np thanks @dboslee

**Comment by arkodg** [comment]
@dboslee checking in to see if you are still actively working on this PR

**Comment by arkodg** [comment]
closing in favor of https://github.com/envoyproxy/gateway/pull/3271

**Pull request closed without merging**